go/enclave/enclave_test.go

```diff
@@ -0,0 +1,225 @@
+	if err != nil {
+		t.Error(err)
+	}
+
+	// callMsg serialized as a Hex
+	callMsgHex := "0x" + gethcommon.Bytes2Hex(callMsgBytes)
+	if err != nil {
+		panic(err)
+	}
+
+	// callMsg encrypted with the VK
+	encryptedParams, err := ecies.Encrypt(rand.Reader, _enclavePubKey, []byte("[\""+callMsgHex+"\"]"), nil, nil)
+	if err != nil {
+		t.Errorf("could not encrypt the following request params with enclave public key: %s. Cause: %s", []byte{123}, err)
+	}
+
+	// Run gas Estimation
+	_, err = enclave.EstimateGas(encryptedParams)
+	if err == nil {
+		t.Fatal("enclave does not have the viewing key to successfully encrypt")
+	}
+}
+
+func gasEstimateNoCallMsgFrom(t *testing.T, _ wallet.Wallet, enclave common.Enclave, _ *rpc.ViewingKey) {
+	// create the callMsg
+	callMsg := datagenerator.CreateCallMsg()
+	callMsgBytes, err := json.Marshal(callMsg)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// callMsg serialized as a Hex
+	callMsgHex := "0x" + gethcommon.Bytes2Hex(callMsgBytes)
+	if err != nil {
+		panic(err)
+	}
+
+	// callMsg encrypted with the VK
+	encryptedParams, err := ecies.Encrypt(rand.Reader, _enclavePubKey, []byte("[\""+callMsgHex+"\"]"), nil, nil)
+	if err != nil {
+		t.Errorf("could not encrypt the following request params with enclave public key: %s. Cause: %s", []byte{123}, err)
+	}
+
+	// Run gas Estimation
+	_, err = enclave.EstimateGas(encryptedParams)
+	if err == nil {
+		t.Fatal("enclave does not have the viewing key to successfully encrypt")
+	}
+}
+
+func gasEstimateInvalidCallMsg(t *testing.T, _ wallet.Wallet, enclave common.Enclave, _ *rpc.ViewingKey) {
+	// create a L2Tx instead
+	callMsg := datagenerator.CreateL2Tx()
+	callMsgBytes, err := json.Marshal(callMsg)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// callMsg serialized as a Hex
+	callMsgHex := "0x" + gethcommon.Bytes2Hex(callMsgBytes)
+	if err != nil {
+		panic(err)
+	}
+
+	// callMsg encrypted with the VK
+	encryptedParams, err := ecies.Encrypt(rand.Reader, _enclavePubKey, []byte("[\""+callMsgHex+"\"]"), nil, nil)
+	if err != nil {
+		t.Errorf("could not encrypt the following request params with enclave public key: %s. Cause: %s", []byte{123}, err)
+	}
+
+	// Run gas Estimation
+	_, err = enclave.EstimateGas(encryptedParams)
+	if err == nil {
+		t.Fatal("enclave should not parse invalid params")
+	}
+}
+
+// registerWalletViewingKey takes a wallet and registers a VK with the enclave
+func registerWalletViewingKey(t *testing.T, enclave common.Enclave, w wallet.Wallet) (*rpc.ViewingKey, error) {
+	// generate the VK from the wallet
+	key, err := rpc.GenerateAndSignViewingKey(w)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// encrypt the VK
+	encryptedViewingKeyBytes, err := ecies.Encrypt(rand.Reader, _enclavePubKey, key.PublicKey, nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// add the VK to the enclave
+	return key, enclave.AddViewingKey(encryptedViewingKeyBytes, key.SignedKey)
+}
+
+// createRandomEnclave returns a new instance of the enclave with random values
```

Comment:
Maybe call this stuff 'arbitrary' rather than 'random'. I was getting ready for a rant about deterministic tests haha.

Reply:
To be fair the values are not random at all ! I just noticed that, but let me change that name, you're right.